modules/ds18b20/ds18b20.h

```diff
@@ -0,0 +1,315 @@
+    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING 
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+    OTHER DEALINGS IN THE SOFTWARE.
+   ----------------------------------------------------------------------
+\endverbatim
+ */
+#ifndef DS18B20_H
+#define DS18B20_H
+
+/**
+ * @addtogroup STM32Fxxx_HAL_Libraries
+ * @{
+ */
+
+/**
+ * @defgroup DS12820
+ * @brief    Library for interfacing DS18B20 temperature sensor from Dallas semiconductors - http://stm32f4-discovery.net/2015/07/hal-library-06-ds18b20-for-stm32fxxx/
+ * @{
+ *
+ * With this you can read temperature, set and get temperature resolution from 9 to 12 bits and check if device is DS18B20.
+ * 
+ * Pin for STM32Fxxx is the same as set with @ref OneWireStruct library.
+ *
+ * \par Changelog
+ *
+\verbatim
+ Version 1.0
+  - First release
+\endverbatim
+ *
+ * \par Dependencies
+ *
+\verbatim
+ - STM32Fxxx HAL
+ - TM OneWireStruct
+ - TM GPIO
+ - defines.h
+\endverbatim
+ */
+
+#include <modules/onewire/onewire.h>
+
+/* OneWireStruct version check */
+// #if ONEWIRE_H < 100
+// #error "Please update TM OneWireStruct LIB, minimum required version is 2.0.0. Download available on stm32f4-discovery.net website"
+// #endif
+
+/**
+ * @defgroup DS18B20_Macros
+ * @brief    Library defines
+ * @{
+ */
+
+/* More details: https://datasheets.maximintegrated.com/en/ds/DS18B20.pdf */
+
+/* Every OneWireStruct chip has different ROM code, but all the same chips has same family code */
+/* in case of DS18B20 this is 0x28 and this is first byte of ROM address */
+#define DS18B20_FAMILY_CODE				0x28
+
+/* DS18B20 Commands */
+#define DS18B20_CMD_ALARMSEARCH			0xEC
+#define DS18B20_CMD_CONVERTTEMP         0x44    /* Convert temperature */
+
+/* DS18B20 Resolution defines */
+#define DS18B20_DECIMAL_STEPS_12BIT		0.0625
+#define DS18B20_DECIMAL_STEPS_11BIT		0.125
+#define DS18B20_DECIMAL_STEPS_10BIT		0.25
+#define DS18B20_DECIMAL_STEPS_9BIT		0.5
+
+/* Bits locations for resolution */
+#define DS18B20_RESOLUTION_R1           6
+#define DS18B20_RESOLUTION_R0           5
+
+/* CRC enabled */
+#ifdef DS18B20_USE_CRC  
+#define DS18B20_DATA_LEN                9
+#else
+#define DS18B20_DATA_LEN                2
+#endif
+
+/* DS18B20 Common Numbers */
+#define DS18B20_DATA_CRC_BYTE  8
+#define DS18B20_DATA_LSB       0
+#define DS18B20_DATA_MSB       1
+
+#define DS18B20_MAX_TEMP_DEG_C 125
+#define DS18B20_MIN_TEMP_DEG_C -55
+
+#define DS18B20_CONFIG_REGISTER_BYTE          4
+#define DS18B20_CONFIG_REGISTER_R0_R1_BITMASK 0x60
+#define DS18B20_CONFIG_REGISTER_RESERVED_BITS 5
+
+#define DS18B20_TEMP_SIGN_BITMASK             0x8000
+
+/* Return Codes*/
+#define DS18B20_USAGE_ERROR           -1
+#define DEVICE_NOT_DS18B20             0
+#define DS18B20_CONVERSION_SUCCESS     1
+#define DS18B20_CONVERSION_IN_PROGRESS 2
+#define DS18B20_CRC_INVALID            3

```

Comment:
It looks like you're using negative values to denote error conditions.  Is `DEVICE_NOT_DS18B20` considered an error?  Should this be a negative value?

Reply:
#define is acceptable here.  But any reason these couldn't be an enum?

Reply:
These can be converted to enum, and are in the process of another refactor to accomodate for this.

Reply:
It does make sense to take the other error conditions and move them to negative. I will do this.

Reply:
Please correct me if I'm wrong but I think you need to update the rest of your comments below with these return codes

Reply:
The biggeest reason for having error conditions be negative is because it's easy to then check for errors (or success):
if (status < 0)